Return contiguous/non-contiguous flag in statusQuery

darrell-k opened the pull request:
To allow proper handling of albums with non-contiguous track groups in the "album style" play queue (Material only for now, but there's always hope...)

It's all conditioned by the /2/ tag.

I've coded the corresponding Material changes and will submit a PR for that when we've ironed out the wrinkles in this one.

darrell-k committed:
Return contiguous/non-contiguous flag in statusQuery

Signed-off-by: darrell-k <darrell@darrell.org.uk>

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4281,6 +4285,25 @@ sub statusQuery {
+
+						if ( $tags =~ /2/ ) {
+							# build a hash containing the grouping data for each album instance/track in the play queue.
+							my $track = $request->{_results}->{'playlist_loop'}[$count];
+							my $album = $track->{'album_id'};
+							my $tracknum = $track->{'tracknum'};
+							my $albumTrack = $album*10000+$tracknum; # used to detect a new album or a repeat of tracks from the same album
+							my $work = $track->{'work_id'} || '';
+							my $grouping = $track->{'grouping'} || '';
+							my $performance = $track->{'performance'} || '';
+							my $group = "$work##$grouping##$performance";
```
Please do not define variables which are only used once. Just access the raw value when needed. IMHO it's harder to read lengthy code caused by variable assignments than reading slightly less visually pleasing expressions with lengthy value references.

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4281,6 +4285,25 @@ sub statusQuery {
 									'playlist index', $idx, $fast
 								);
+
+						if ( $tags =~ /2/ ) {
```
Did you ever document the tag `2`? I didn't find it in the CLI reference.

darrell-k replied:
It's documented in the albums section, though that needs updating to mention it now also returns the `contiguous_groups` flag. I'll do this. I'll also add it to the compound queries/status section (it's not a songinfo tag, as it needs a view of the whole play queue to do its stuff).

michaelherger replied:
Please double check. I can't find any hint about it. Thanks!

darrell-k replied:
![image](https://github.com/user-attachments/assets/d5abf66c-0401-4c55-bd0f-3c9d9f7bed08)

michaelherger replied:
Got it, thanks!

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4281,6 +4285,25 @@ sub statusQuery {
+							}
+							$lastAlbum = $album;
+							$lastAlbumTrack = $albumTrack;
+							$request->{_results}->{'playlist_loop'}[$count]->{'albumNumber'} = $albumNumber;
```
Oooh... don't mess with a request's inner data structures. Store the data elsewhere if needed. But don't modify an object's internal data.

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4291,6 +4314,32 @@ sub statusQuery {
 
+				if ( $tags =~ /2/ ) {
+					# process each album group in the playlist
+					foreach my $albumGroup (keys %groups) {
```
Are you expecting any particular order here? `keys` doesn't have one, AFAIK, even if they are just numbers.

darrell-k replied:
Doesn't matter, we just need to process all the groups.

michaelherger reviewed (changes requested):
Please see comments. This needs more work. I can't process all those helper variables...

Is there a relevant forum thread somewhere?

darrell-k commented:
> Please see comments. This needs more work. I can't process all those helper variables...
> 
> Is there a relevant forum thread somewhere?

About to push changes which hopefully address the issues. There's no forum thread, I noticed myself that the Material Play Queue was all messed up in "album style" mode (ie "toggle view" in the playlist options) when displaying an album with non-contiguous groups of tracks. As I said at the top, I've got some Material changes to submit to Craig to fix this, once this PR is merged.

darrell-k committed:
changes after review

Signed-off-by: darrell-k <darrell@darrell.org.uk>

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4291,6 +4311,31 @@ sub statusQuery {
+						my $groupSeen = {};
+						# determine whether album group contains contiguous or non-contiguous groups of tracks
+						foreach my $track ( sort { $a <=> $b } keys %{$groups{$albumGroup}} ) {
+							my $thisTrackGroup = $groups{$albumGroup}{$track};
```
This should be `$group{$albumGroup}->{$track}`, right (the arrow is missing)?

darrell-k replied:
It appears to work?

michaelherger replied:
Would my version work, too? I believe some Perl versions were tolerant towards this use of hash refs, but later versions would complain. Please prove me wrong again - or use the arrow notation. It's a hash ref, not a hash.

darrell-k replied:
You're almost certainly correct, I'll make the change and test. I still get myself into a mess with hash/hash ref questions.

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4291,6 +4311,31 @@ sub statusQuery {
+								}
+							}
+						}
+						# now set the contiguous_groups flag in the playlist_loop array.
```
`playlist_loop` would only exist in menu mode. In CLI mode this would be `item_loop`. See line 4163. Don't know whether `tag:2` is relevant in non menu mode, but let's make sure we handle this correctly.

darrell-k replied:
It's the other way around, I think?

darrell-k replied:
```
		if ( $menuMode ) {
			# Set required tags for menuMode
			$tags = 'aAlKNcxJ';
		}

```
So tags =~ /2/ is enough, I think.

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4291,6 +4311,31 @@ sub statusQuery {
+							}
+						}
+						# now set the contiguous_groups flag in the playlist_loop array.
+						foreach ( grep { $plIndexAlbumSequence{$_->{'playlist index'}} == $albumGroup } @{ $request->getResult('playlist_loop') }) {
+							$_->{'contiguous_groups'} = $nonContiguous ? 0 : 1;
+						}
```
It's my understanding that in this loop you're looking for the playlist item with the given index, to set the `contiguous_groups` flag. In order to do so you loop through the list for each item in another list, creating nested loops. That's potentially very inefficient (eg. for a long playlist). I'd recommend you create a hash keyed by the index, referencing the playlist items _outside_ the loop. You only have to do this once:

```perl
my %playlistItems = map { $_->{'playlist index'} => $_ } @{ $request->getResult('playlist_loop'} };
``` 

Then this loop becomes a simple conditional statement:

```perl
if (my $playlistItem = $playlistItems{$albumGroup}) {
   $playlistItem->{'contiguous_groups'} = $nonContiguous ? 0 : 1;
}
```

Unless, of course, I'm wrong once again. I'm simply trying to avoid nested loops.

darrell-k replied:
`playlist index` is not `$albumGroup`. There would usually be many playlist entries for each album group, so we need to loop through playlist_loop, one way or the other.

michaelherger replied:
You're iterating over all items for which the `playlist index` would be equal to `$albumGroup`. This can only ever be one single item, right, because that value is the position within the playlist, isn't it?

michaelherger replied:
Oh, there's one more level of indirection... `$plIndexAlbumSequence`. There's always something more...

darrell-k replied:
> You're iterating over all items for which the `playlist index` would be equal to `$albumGroup`. This can only ever be one single item, right, because that value is the position within the playlist, isn't it?

No, could (almost certainly will) be multiple items. 

It's like this:

```
Album Group    Playlist index
1                          0
1                          1
1                          2
2                          3
2                          4
2                          5

```
So for each album group, we loop through the items to see if the tracks are contiguous wrt work/grouping/performance. Then we set the contiguous_groups flag for every track in playlist_loop accordingly, which needs another loop.

michaelherger reviewed (changes requested):
I don't like nested loops, as they're usually wasteful. Please see below, whether my suggestion would work. Thanks!

darrell-k committed:
correct hashref syntax

Signed-off-by: darrell-k <darrell@darrell.org.uk>

michaelherger commented on Slim/Control/Queries.pm:
```diff
@@ -4281,6 +4286,21 @@ sub statusQuery {
+							my $track = @{ $request->getResult('playlist_loop') }[$count];
+							my $albumTrack = $track->{'album_id'}*10000+$track->{'tracknum'}; # used to detect a new album or a repeat of tracks from the same album
+							my $group = "$track->{'work_id'}##$track->{'grouping'}##$track->{'performance'}";
+							if ( $track->{'album_id'} != $lastAlbum || $albumTrack < $lastAlbumTrack ) {
```
Why would you only check whether the `$albumTrack < $lastAlbumTrack`? That value isn't always in ascending order, is it? Its highest order is by album ID - which can be randomly distributed in the queue.

darrell-k replied:
Well it's also checking $track->{'album_id'} != $lastAlbum so a play queue like:

album 1 track 1
album 1 track 2
album 1 track 3
album 1 track 2
album 2 track 1
album 2 track 2

will be split into 3 groups, which is what we want.

michaelherger replied:
Is an album in shuffled order not considered contiguous?

darrell-k replied:
It would much more likely have tracks flagged as contiguous, because a new group would be created every time the track number jumped backwards within an album. But Material would still create a new subheading when the track group (ie work/grouping/performance) changed.

This PR is about giving Material the data to _suppress_ those subheadings when using the "album style" (hierarchical) play queue view where there are non-contiguous track groupings.

In so much as the "album style" view is useful at all when shuffling the playlist, it seems to behave, but I'll do some more testing.

darrell-k replied:
Repeatedly jabbing the shuffle button on a variety of play queues hasn't revealed anything bad, so IMO this is ready for merge.

michaelherger committed:
Some food for thought: simplify how we evaluate `contiguous_groups` for the queue.

I didn't like how you iterated over the full playlist again and again for each group in order to set the `contiguous_groups`. So I was wondering whether this couldn't be simplified. This was more work than I thought, but I believe it _is_ simpler, uses less helper variables etc.

While we build the tracks list, enrich the track information with the grouping information previously stored in `%plIndexAlbumSequence` and `%groups`. Keep the latter around to store references to the track object grouped by album group.

In the second step use `%groups` to directly set the `contiguous_groups` flag on the track objects. No need to iterate the track list repeatedly just to find whether they are part of the current group, as we only process the tracks of the current group. That way we should only process each track once, instead of all of them for each group.

My testing also showed that you were expecting a lot of data to be around, without verifying it. Just sending the `tag:2` would not have provided track number, works etc. So I added a few lines to make sure we have what we need.

Unfortunately I don't have usable test data for this feature, or I maybe don't understand the use case anyway. So please don't trust me blindly, but give this some good testing before you pick what you can use (or not).

Signed-off-by: Michael Herger <michael@herger.net>

darrell-k committed:
Merge pull request #16 from michaelherger/works-subheadings-titles

Food for thought: simplify how we evaluate `contiguous_groups` for the queue

darrell-k committed:
tweaks to Michael's rework

Signed-off-by: darrell-k <darrell@darrell.org.uk>
